ddl/ddl.go

```diff
@@ -1291,6 +1321,82 @@ func CancelJobs(txn kv.Transaction, ids []int64) ([]error, error) {
 	return errs, nil
 }
 
+// cancelConcurrencyJobs cancels the DDL jobs that are in the concurrent state.
+func cancelConcurrencyJobs(se sessionctx.Context, ids []int64) ([]error, error) {
+	failpoint.Inject("mockCancelConcurencyDDL", func(val failpoint.Value) {
+		if val.(bool) {
+			failpoint.Return(nil, errors.New("mock commit error"))
+		}
+	})
+	if len(ids) == 0 {
+		return nil, nil
+	}
+	var jobSet = make(map[int64]int) // jobID -> error index
+
+	sess := newSession(se)
+	err := sess.begin()
+	if err != nil {
+		return nil, err
+	}
+
+	idsStr := make([]string, 0, len(ids))
+	for idx, id := range ids {
+		jobSet[id] = idx
+		idsStr = append(idsStr, strconv.FormatInt(id, 10))
+	}
+
+	jobs, err := getJobsBySQL(sess, JobTable, fmt.Sprintf("job_id in (%s) order by job_id", strings.Join(idsStr, ", ")))
+	if err != nil {
+		sess.rollback()
+		return nil, err
+	}
+
+	errs := make([]error, len(ids))
+
+	for _, job := range jobs {
+		i, ok := jobSet[job.ID]
+		if !ok {
+			continue
+		}
+		delete(jobSet, job.ID)
+		if job.IsDone() || job.IsSynced() {
+			errs[i] = dbterror.ErrCancelFinishedDDLJob.GenWithStackByArgs(job.ID)
+			continue
+		}
+		// If the state is rolling back, it means the work is cleaning the data after cancelling the job.
+		if job.IsCancelled() {
+			errs[i] = dbterror.ErrCancelledDDLJob.GenWithStackByArgs(job.ID)
+			continue
+		}
+		if job.IsRollingback() || job.IsRollbackDone() {
+			continue
+		}
+		if !job.IsRollbackable() {
+			errs[i] = dbterror.ErrCannotCancelDDLJob.GenWithStackByArgs(job.ID)
+			continue
+		}
+		job.State = model.JobStateCancelling
+		// Make sure RawArgs isn't overwritten.
+		err := json.Unmarshal(job.RawArgs, &job.Args)
+		if err != nil {
+			errs[i] = errors.Trace(err)
+			continue
+		}
```

Comment:
It seems the Args encoding is not necessary.

Reply:
I think we can change the argument of `updateRawArgs` in updateDDLJob2Table to `false`. Then we can remove it. 
Besides, to make sure `RawArgs` is not changed before and after cancel, we can add a test

Reply:
yes, I agree, I will do it in another PR